### PR TITLE
Added XTags

### PR DESCRIPTION
**Pull request by RohanGoyalDev**
A possible replacement for  [Spigot's Tag system](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/Tag.html) that can be used for 1.8 and can have XSound, XBlock, XParticle genericised later on to ease in finding enum values that all carry a common property

### CODE REVIEW

**Comment by RohanGoyalDev** [comment]
I'll add back generics later, didn't see he removed them

File: src/main/java/com/cryptomorin/xseries/XMaterial.java [hunk 1]
```diff
@@ -2071,6 +2071,21 @@ public boolean isSupported() {
...
+    /**
+     * Returns a set of all supported materials out of the set provided
+     * <p>
+     *
+     * @param materials {@link Set} of the materials to be filtered
+     *
+     * @return {@link Set} of materials that contains only the supported materials out of the provided {@link Set}
+     */
+
+    public static Set<XMaterial> filterSupported(Set<XMaterial> materials) {
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XMaterial.java:2083]: Why is this useful?

File: src/test/XTagTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  @Test
+  void checkNonInheritTag() {
+    final XMaterial target = XMaterial.TUBE_CORAL;
+
+    assertTrue(XTag.CORAL_PLANTS.isTagged(target));
+  }
+
+  @Test
+  void checkInheritTag() {
```
Comment (CryptoMorin) [inline src/test/XTagTest.java:20]: You should add all tests to `Tests` class. I personally don't see the point of separating these.

File: src/main/java/com/cryptomorin/xseries/XTag.java [hunk 3]
```diff
@@ -0,0 +1,2221 @@
+package com.cryptomorin.xseries;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public final class XTag {
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XTag.java:10]: Wouldn't it make more sense to use an enum to avoid this mess?

**Review by CryptoMorin — CHANGES_REQUESTED** [review]

**Comment by CryptoMorin** [comment]
I'm personally not that familiar with tags, but have you looked at [XMaterialUtil](https://github.com/CryptoMorin/XSeries/blob/master/src/main/java/com/cryptomorin/xseries/XMaterialUtil.java)? Should we remove that and merge it with XTag?

> can have XSound, XBlock, XParticle genericised later on to ease in finding enum values that all carry a common property
Also, if this means changing those classes, discuss this before making any changes.

File: src/main/java/com/cryptomorin/xseries/XTag.java (hunk 3 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XTag.java:10]: Also, using static imports sounds reasonable at this point.

File: src/main/java/com/cryptomorin/xseries/XTag.java [hunk 4]
```diff
@@ -0,0 +1,2221 @@
...
+        MUSHROOM_GROW_BLOCK.inheritFrom(NYLIUM);
+        MOSS_REPLACEABLE = new XTag();
+        MOSS_REPLACEABLE.inheritFrom(CAVE_VINES,
+                DIRT,
+                BASE_STONE_OVERWORLD);
+    }
+
+    private XTag(@NonNull XMaterial... values) {
+        this.values =
+            Collections.unmodifiableSet(new HashSet<>(Arrays.asList(values)));
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XTag.java:2192]: Use an EnumSet

**Comment by RohanGoyalDev** [comment]
> I'm personally not that familiar with tags, but have you looked at [XMaterialUtil](https://github.com/CryptoMorin/XSeries/blob/master/src/main/java/com/cryptomorin/xseries/XMaterialUtil.java)? Should we remove that and merge it with XTag?
> 
> > can have XSound, XBlock, XParticle genericised later on to ease in finding enum values that all carry a common property
> 
> Also, if this means changing those classes, discuss this before making any changes.

Anything that is an enum can be used under XTag, so XBlock and XParticle can't rn but XPotion, XSound, XBiome, XEnchantment can without any changes in those enums

**Comment by CryptoMorin** [comment]
I think you should really fix the enum issue before adding more stuff.

**Comment by RohanGoyalDev** [comment]
> I think you should really fix the enum issue before adding more stuff.

Can you gimme like an example or class hierarchy? I thought of many ways to use an enum for this but all of them still create a mess

**Comment by CryptoMorin** [comment]
Hmm, yeah you're right, after trying a few methods myself there's no way to make this cleaner while keeping the type safety or the cleanness from an API perspective.

**Comment by RohanGoyalDev** [comment]
I will keep adding more Tags. If you or anyone else has a good idea on cleanup gimme an example and ill refactor the code accordingly if that works

**Comment by CryptoMorin** [comment]
I think we should support tagged mappings too.
isTagged will still return a boolean without being affected, but we can add a new method of(key)
The only tradeoff is the performance and memory of `EnumMap` which will be extremely memory inefficient for a class like XMaterial because it has a lot of enums since `EnumMap` doesn't do Jumbo optimizations.

We can fix that if we make our own wrapper.
https://pastebin.com/BQEBypwb

Btw while testing this I noticed you did this:
```java
if (this.values.isEmpty()) newValues = EnumSet.copyOf((EnumSet<T>) this.values);
else newValues = EnumSet.copyOf(this.values);
```
There's no difference between these two lines, was this intentional?



We could also take Spigot's approach and use a switch statement in a simple static method instead of a map.

**Comment by RohanGoyalDev** [comment]
> ```java
> if (this.values.isEmpty()) newValues = EnumSet.copyOf((EnumSet<T>) this.values);
> else newValues = EnumSet.copyOf(this.values);
> ```


Well those lines are intentional because of the illegal argument that `EnumSet.copyOf()` throws when the `Collection` passed in as a parameter is empty. And there are Tags like `ITEMS_FURNACE_MATERIALS` which are for some apparent reason left empty in spigot 1.18. To account for those and the tags that just inherit values from other tags and have no seperate value I had to made a seperate constructor because `T.class` is not a thing and `EnumSet.noneOf()` needs that, if i try to solve both of those issues then I cant make the `values` class variable an unmodifiable set because `Collections.unModifiableSet()` returns a set that cant be casted to `EnumSet`

**Comment by RohanGoyalDev** [comment]
> > ```java
> > if (this.values.isEmpty()) newValues = EnumSet.copyOf((EnumSet<T>) this.values);
> > else newValues = EnumSet.copyOf(this.values);
> > ```
> 
> Well those lines are intentional because of the illegal argument exception that `EnumSet.copyOf()` throws when the `Collection` passed in as a parameter is empty UNLESS its another `EnumSet`. And there are Tags like `ITEMS_FURNACE_MATERIALS` which are for some apparent reason left empty in spigot 1.18. To account for those and the tags that just inherit values from other tags and have no seperate value I had to made a seperate constructor because `T.class` is not a thing and `EnumSet.noneOf()` needs that, if i try to solve both of those issues then I cant make the `values` class variable an unmodifiable set because `Collections.unModifiableSet()` returns a set that cant be casted to `EnumSet`

**Comment by CryptoMorin** [comment]
If you're done with this, tell me to merge it.

**Comment by RohanGoyalDev** [comment]
I think its done for now apart from the tagged mappings. I'll keep adding more tags which i see useful in more pull requests later on

**Pull request merged**